src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -1,16 +1,935 @@
+using System.Collections.Immutable;
 using System.Linq;
+using System.Text;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using Xunit;
-
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
 {
+    using static Instruction;
+    internal enum Instruction
+    {
+        Write,
+        Yield,
+        AwaitSlow,
+        AwaitFast,
+        YieldBreak
+    }
+
     [CompilerTrait(CompilerFeature.AsyncStreams)]
     public class CodeGenAsyncIteratorTests : EmitMetadataTestBase
     {
+        // PROTOTYPE(async-streams)
+        // Test missing types/members
+        // Test with yield or await in try/catch/finally
+        // More tests with exception thrown
+        // Test yield break
+        // There is a case in GetIteratorElementType with IsDirectlyInIterator that relates to speculation, needs testing
+        // yield break disallowed in finally and top-level script (see BindYieldBreakStatement); same for yield return (see BindYieldReturnStatement)
+        // test local function
+        // binding for yield return (BindYieldReturnStatement) validates escape rules, needs testing
+        // test yield in async lambda (still error)
+        // test exception handling (should capture and return the exception via the promise)
+        // test IAsyncEnumerable<U> M<U>() ...
+        // test missing AsyncTaskMethodBuilder<T> or missing members Create(), Task, ...
+        // test with IAsyncEnumerable<dynamic>
+        // other tests with dynamic?
+        // test should cover both case with AwaitOnCompleted and AwaitUnsafeOnCompleted
+        // test `async IAsyncEnumerable<int> M() { return TaskLike(); }`
+        // Can we avoid making IAsyncEnumerable<T> special from the start? Making mark it with an attribute like we did for task-like?
+        // Test normal break in async iterator
+        // WaitForNextAsync is resilient to be called out of turn. Test that.
+        // Test a plain return statement
+
+        [Fact]
+        public void AsyncIteratorWithAwaitCompletedAndYield()
+        {
+            string source = @"
+using static System.Console;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        Write(""1 "");
+        await System.Threading.Tasks.Task.CompletedTask;
+        Write(""2 "");
+        yield return 3;
+        Write("" 4 "");
+    }
+    static async System.Threading.Tasks.Task Main()
+    {
+        Write(""0 "");
+        foreach await (var i in M())
+        {
+            Write(i);
+        }
+        Write(""5"");
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(source + s_common, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "0 1 2 3 4 5", verify: Verification.Passes);
+
+            verifier.VerifyIL("C.M", @"
```

Comment:
Here's the decompiled code for this scenario (generated via `verifier.Dump()`), with a few comments added:

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.ExceptionServices;
using System.Threading.Tasks;
using System.Threading.Tasks.Sources;

[assembly: CompilationRelaxations(8)]
[assembly: RuntimeCompatibility(WrapNonExceptionThrows = true)]
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.Default | DebuggableAttribute.DebuggingModes.DisableOptimizations | DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints | DebuggableAttribute.DebuggingModes.EnableEditAndContinue)]
[assembly: AssemblyVersion("0.0.0.0")]
internal class C
{
	// This contains the lowered code for method `M()`
	[CompilerGenerated]
	private sealed class <M>d__0 : IAsyncEnumerable<int>, IAsyncEnumerator<int>, IAsyncDisposable, IValueTaskSource<bool>, IStrongBox<ManualResetValueTaskSourceLogic<bool>>, IAsyncStateMachine
	{
		public int <>1__state;
		public AsyncVoidMethodBuilder <>t__builder;
		private TaskAwaiter <>u__1;

		private int <>2__current;
		public ManualResetValueTaskSourceLogic<bool> <>v__promiseOfValueOrEnd;
		public bool <>w__promiseIsActive;

		ManualResetValueTaskSourceLogic<bool> IStrongBox<ManualResetValueTaskSourceLogic<bool>>.Value
		{
			[DebuggerHidden]
			get { return ref <>v__promiseOfValueOrEnd; }
		}

		private void MoveNext()
		{
			int num = <>1__state;
			try
			{
				TaskAwaiter awaiter;
				int <>1__state2;
				switch (num)
				{
				default:
					Console.Write("1 ");
					awaiter = Task.CompletedTask.GetAwaiter();
					if (!awaiter.IsCompleted)
					{
						num = (<>1__state = 0);
						<>u__1 = awaiter;

// this is the handshake change to activate and reset the promise when starting background execution
						if (!<>w__promiseIsActive)
						{
							<>w__promiseIsActive = true;
							<>v__promiseOfValueOrEnd.Reset();
						}

						<M>d__0 <M>d__ = this;
						<>t__builder.AwaitUnsafeOnCompleted(ref awaiter, ref <M>d__);
						break;
					}
					goto IL_0097;
				case 0:
					awaiter = <>u__1;
					<>u__1 = default(TaskAwaiter);
					num = (<>1__state = -1);
					goto IL_0097;
				case 1:
					{
						Console.Write(" 4 ");

// reached the end
						if (!<>w__promiseIsActive)
						{
							<>w__promiseIsActive = true;
							<>v__promiseOfValueOrEnd.Reset();
						}
						<>v__promiseOfValueOrEnd.SetResult(false);

						break;
					}
					IL_0097:
					awaiter.GetResult();
					Console.Write("2 ");

// reached `yield return 3;`
					<>2__current = 3;
					<>1__state2 = <>1__state;
					<>1__state = 1;
					if (<>w__promiseIsActive)
					{
						<>v__promiseOfValueOrEnd.SetResult(true);
					}
					break;
				}
			}
			catch (Exception exception)
			{
				<>1__state = -2;
				if (<>w__promiseIsActive)
				{
					<>v__promiseOfValueOrEnd.SetException(exception);
					goto end_IL_010b;
				}
				throw;
				end_IL_010b:;
			}
		}

		void IAsyncStateMachine.MoveNext()
		{
			//ILSpy generated this explicit interface implementation from .override directive in MoveNext
			this.MoveNext();
		}

		[DebuggerHidden]
		private void SetStateMachine(IAsyncStateMachine stateMachine) { }

		void IAsyncStateMachine.SetStateMachine(IAsyncStateMachine stateMachine)
		{
			//ILSpy generated this explicit interface implementation from .override directive in SetStateMachine
			this.SetStateMachine(stateMachine);
		}

		[DebuggerHidden]
		IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator()
		{
			return this;
		}

		[DebuggerHidden]
		ValueTask<bool> IAsyncEnumerator<int>.WaitForNextAsync()
		{
			if (<>1__state == -2)
			{
				return default(ValueTask<bool>);
			}
			if (!<>w__promiseIsActive || <>1__state == -1)
			{
				<M>d__0 <M>d__ = this;
				<>t__builder.Start(ref <M>d__);
			}
			return new ValueTask<bool>(this, <>v__promiseOfValueOrEnd.Version);
		}

		[DebuggerHidden]
		int IAsyncEnumerator<int>.TryGetNext(out bool success)
		{
			if (<>w__promiseIsActive)
			{
				<>w__promiseIsActive = false;
			}
			else
			{
				<M>d__0 <M>d__ = this;
				<>t__builder.Start(ref <M>d__);
			}
			if (!<>w__promiseIsActive && <>1__state != -2)
			{
				success = true;
				return <>2__current;
			}
			success = false;
			return 0;
		}

#region Implementation of promise
		[DebuggerHidden]
		bool IValueTaskSource<bool>.GetResult(short token)
		{
			return <>v__promiseOfValueOrEnd.GetResult(token);
		}

		[DebuggerHidden]
		ValueTaskSourceStatus IValueTaskSource<bool>.GetStatus(short token)
		{
			return <>v__promiseOfValueOrEnd.GetStatus(token);
		}

		[DebuggerHidden]
		void IValueTaskSource<bool>.OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
		{
			<>v__promiseOfValueOrEnd.OnCompleted(continuation, state, token, flags);
		}
#endregion

		[DebuggerHidden]
		ValueTask IAsyncDisposable.DisposeAsync()
		{
			<>v__promiseOfValueOrEnd.Reset();
			<>1__state = -1;
			return default(ValueTask);
		}
	}

	// This contains the lowered code for method `Main()`
	[CompilerGenerated]
	private sealed class <Main>d__1 : IAsyncStateMachine
	{
		public int <>1__state;

		public AsyncTaskMethodBuilder <>t__builder;

		private IAsyncEnumerator<int> <>s__1;

		private object <>s__2;

		private int <>s__3;

		private int <i>5__4;

		private bool <>s__5;

		private ValueTaskAwaiter<bool> <>u__1;

		private ValueTaskAwaiter <>u__2;

		private void MoveNext()
		{
			int num = <>1__state;
			try
			{
				ValueTaskAwaiter awaiter;
				object obj2;
				switch (num)
				{
				default:
					Console.Write("0 ");
					<>s__1 = M().GetAsyncEnumerator();
					<>s__2 = null;
					<>s__3 = 0;
					goto case 0;
				case 0:
				{
					<Main>d__1 <Main>d__;
					try
					{
						if (num != 0)
						{
							goto IL_0075;
						}
						ValueTaskAwaiter<bool> awaiter2 = <>u__1;
						<>u__1 = default(ValueTaskAwaiter<bool>);
						num = (<>1__state = -1);
						goto IL_00d6;
						IL_0075:
						awaiter2 = <>s__1.WaitForNextAsync().GetAwaiter();
						if (!awaiter2.IsCompleted)
						{
							num = (<>1__state = 0);
							<>u__1 = awaiter2;
							<Main>d__ = this;
							<>t__builder.AwaitUnsafeOnCompleted(ref awaiter2, ref <Main>d__);
							return;
						}
						goto IL_00d6;
						IL_00d6:
						<>s__5 = awaiter2.GetResult();
						if (<>s__5)
						{
							while (true)
							{
								<i>5__4 = <>s__1.TryGetNext(out bool flag);
								if (flag)
								{
									Console.Write(<i>5__4);
									continue;
								}
								break;
							}
							goto IL_0075;
						}
					}
					catch (object obj)
					{
						obj2 = (<>s__2 = obj);
					}
					if (<>s__1 == null)
					{
						break;
					}
					awaiter = <>s__1.DisposeAsync().GetAwaiter();
					if (!awaiter.IsCompleted)
					{
						num = (<>1__state = 1);
						<>u__2 = awaiter;
						<Main>d__ = this;
						<>t__builder.AwaitUnsafeOnCompleted(ref awaiter, ref <Main>d__);
						return;
					}
					goto IL_0169;
				}
				case 1:
					{
						awaiter = <>u__2;
						<>u__2 = default(ValueTaskAwaiter);
						num = (<>1__state = -1);
						goto IL_0169;
					}
					IL_0169:
					awaiter.GetResult();
					break;
				}
				obj2 = <>s__2;
				if (obj2 != null)
				{
					Exception ex = obj2 as Exception;
					if (ex == null)
					{
						throw obj2;
					}
					ExceptionDispatchInfo.Capture(ex).Throw();
				}
				int <>s__6 = <>s__3;
				<>s__2 = null;
				<>s__1 = null;
				Console.Write("5");
			}
			catch (Exception exception)
			{
				<>1__state = -2;
				<>t__builder.SetException(exception);
				return;
			}
			<>1__state = -2;
			<>t__builder.SetResult();
		}

		void IAsyncStateMachine.MoveNext()
		{
			//ILSpy generated this explicit interface implementation from .override directive in MoveNext
			this.MoveNext();
		}

		[DebuggerHidden]
		private void SetStateMachine(IAsyncStateMachine stateMachine)
		{
		}

		void IAsyncStateMachine.SetStateMachine(IAsyncStateMachine stateMachine)
		{
			//ILSpy generated this explicit interface implementation from .override directive in SetStateMachine
			this.SetStateMachine(stateMachine);
		}
	}

	[AsyncStateMachine(typeof(<M>d__0))]
	[DebuggerStepThrough]
	private static IAsyncEnumerable<int> M()
	{
		<M>d__0 <M>d__ = new <M>d__0();
		<M>d__.<>t__builder = AsyncVoidMethodBuilder.Create();
		<M>d__.<>1__state = -1;
		AsyncVoidMethodBuilder <>t__builder = <M>d__.<>t__builder;
		<M>d__.<>v__promiseOfValueOrEnd = new ManualResetValueTaskSourceLogic<bool>(<M>d__);
		<M>d__.<>w__promiseIsActive = true;
		return <M>d__;
	}

	[AsyncStateMachine(typeof(<Main>d__1))]
	[DebuggerStepThrough]
	private static Task Main()
	{
		<Main>d__1 <Main>d__ = new <Main>d__1();
		<Main>d__.<>t__builder = AsyncTaskMethodBuilder.Create();
		<Main>d__.<>1__state = -1;
		AsyncTaskMethodBuilder <>t__builder = <Main>d__.<>t__builder;
		<>t__builder.Start(ref <Main>d__); // Note that regular async methods start the state machine running in the background (but async-iterator methods don't)
		return <Main>d__.<>t__builder.Task;
	}

	private static void <Main>()
	{
		Main().GetAwaiter().GetResult();
	}
}

// I've removed common code (interfaces, ManualResetValueTaskSourceLogic, IStrongBox)
```